consensus/tip_selector/tip_selector.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2018 IOTA Stiftung
+ * https://github.com/iotaledger/entangled
+ *
+ * Refer to the LICENSE file for licensing information
+ */
+
+#include <inttypes.h>
+
+#include "consensus/cw_rating_calculator/cw_rating_calculator.h"
+#include "consensus/entry_point_selector/entry_point_selector.h"
+#include "consensus/exit_probability_randomizer/exit_probability_randomizer.h"
+#include "consensus/exit_probability_validator/exit_probability_validator.h"
+#include "consensus/snapshot/snapshot.h"
+#include "consensus/tip_selector/tip_selector.h"
+#include "utils/logger_helper.h"
+
+#define TIP_SELECTOR_LOGGER_ID "consensus_tipselection"
+
+retcode_t iota_consensus_tip_selector_init(
+    tip_selector_t *const tip_selector,
+    cw_rating_calculator_t *const cw_rating_calculator,
+    entry_point_selector_t *const entry_point_selector,
+    ep_randomizer_t *const ep_randomizer,
+    exit_prob_transaction_validator_t *const walker_validator,
+    ledger_validator_t *const ledger_validator,
+    milestone_tracker_t *const milestone_tracker, tangle_t *const tangle,
+    double const alpha, uint32_t const max_txs_below_max_depth,
+    uint32_t const max_depth) {
+  tip_selector->cw_rating_calculator = cw_rating_calculator;
+  tip_selector->entry_point_selector = entry_point_selector;
+  tip_selector->ep_randomizer = ep_randomizer;
+  tip_selector->walker_validator = walker_validator;
+  tip_selector->ledger_validator = ledger_validator;
+  tip_selector->milestone_tracker = milestone_tracker;
+  tip_selector->tangle = tangle;
+  tip_selector->max_txs_below_max_depth = max_txs_below_max_depth;
+  tip_selector->max_depth = max_depth;
+  return RC_OK;
+}
+
+retcode_t iota_consensus_tip_selector_get_transactions_to_approve(
+    tip_selector_t *const tip_selector, size_t const depth,
+    trit_array_t const *const reference, tips_pair *const tips) {
+  retcode_t ret = RC_OK;
+  flex_trit_t ep_trits[FLEX_TRIT_SIZE_243];
+  trit_array_t ep = {.trits = ep_trits,
+                     .num_trits = HASH_LENGTH_TRIT,
+                     .num_bytes = FLEX_TRIT_SIZE_243,
+                     .dynamic = 0};
+  cw_calc_result rating_results = {.cw_ratings = NULL, .tx_to_approvers = NULL};
+  bool consistent = false;
+  hash_stack_t tips_stack = NULL;
+
+  rw_lock_handle_rdlock(
```

Comment:
Once we have a working clone, we should strive to eliminate these locks. They are unnecessary with a slight redesign

Reply:
Agreed